aws/resource_aws_s3_bucket_test.go

```diff
@@ -663,6 +661,162 @@ func TestAccAWSS3Bucket_UpdateAcl(t *testing.T) {
 	})
 }
 
+func TestAccAWSS3Bucket_UpdateGrant(t *testing.T) {
+	resourceName := "aws_s3_bucket.bucket"
+	ri := acctest.RandInt()
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSS3BucketConfigWithGrants(ri),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSS3BucketExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "grant.#", "1"),
+					func(s *terraform.State) error {
+						id := s.RootModule().Resources["data.aws_canonical_user_id.current"].Primary.ID
+						gh := fmt.Sprintf("grant.%v", grantHash(map[string]interface{}{
+							"id":   id,
+							"type": "CanonicalUser",
+							"uri":  "",
+							"permissions": schema.NewSet(
+								schema.HashString,
+								[]interface{}{"FULL_CONTROL", "WRITE"},
+							),
+						}))
+						for _, t := range []resource.TestCheckFunc{
+							resource.TestCheckResourceAttr(resourceName, gh+".permissions.#", "2"),
+							resource.TestCheckResourceAttr(resourceName, gh+".permissions.3535167073", "FULL_CONTROL"),
+							resource.TestCheckResourceAttr(resourceName, gh+".permissions.2319431919", "WRITE"),
+							resource.TestCheckResourceAttr(resourceName, gh+".type", "CanonicalUser"),
+						} {
+							if err := t(s); err != nil {
+								return err
+							}
+						}
+						return nil
+					},
```

Comment:
We prefer to use named functions in the tests, so that it's easier to read.

Reply:
Done